src/Formats/ProtobufSerializer.cpp

```diff
@@ -3445,19 +3444,6 @@ namespace
                 }
             }
 
-            /// Check that we've found matching columns for all the required fields.
```

Comment:
As far as I understand,
- the deleted code "only" performed a safety check, so we could in theory omit it
- the reason it was deleted is that enum `google::protobuf::FileDescriptor` was deprecated in v25.x (users of the API should use "targeted accessors") and ClickHouse handles warnings as errors.
- "required fields" were a Protobuf syntax v2 feature only and removed in v3 (likely they are also not in the new Editions).

What if we retain the deleted code but change it to

```cpp
            if (reader_or_writer.writer)
            {
                for (int i : collections::range(message_descriptor.field_count()))
                {
                    const auto & field_descriptor = *message_descriptor.field(i);
                    if (field_descriptor.is_required() && !field_descriptors_in_use.count(&field_descriptor))
                        throw Exception(ErrorCodes::NO_COLUMN_SERIALIZED_TO_REQUIRED_PROTOBUF_FIELD, "Field {} is required to be set",
                            quoteString(field_descriptor.full_name()));
                }
            }
```

`field_descriptor.is_required()` has internal implementation

```cpp
bool FieldDescriptor::is_required() const {
  return features().field_presence() == FeatureSet::LEGACY_REQUIRED;
}
```

so I assume it will return true for v2 only anyways.